compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -181,6 +187,118 @@ impl VariancedDag {
+                    variances.vars[partition_index.0] = SymbolicVariance::input_external_partition(
+                        partitions.nb_partitions,
+                        partition_index,
+                        max_variance,
+                    );
+                    *(input.variance_mut()) = variances;
+                }
+            }
+        }
+    }
+
+    fn collect_external_input_constraint(&mut self) {
+        for (i, op) in self.dag.operators.clone().iter().enumerate() {
+            if let Operator::Input {
+                out_precision,
+                out_shape,
+            } = op
+            {
+                let partition_index = self.partitions.instrs_partition[i].instruction_partition;
+                if self
+                    .partitions
+                    .p_cut
+                    .is_external_partition(&partition_index)
+                {
+                    let max_variance = &self.partitions.p_cut.external_partitions[self
+                        .partitions
+                        .p_cut
+                        .external_partition_index(partition_index)]
+                    .max_variance;
+
+                    let vars = &self.get_operator(OperatorIndex(i)).variance().vars.clone();
+                    for (i, var) in vars.iter().enumerate() {
+                        if var.coeffs.is_nan() {
+                            assert!(i != partition_index.0);
+                            continue;
+                        }
+                        let constraint = VarianceConstraint {
+                            precision: *out_precision,
+                            partition: partition_index,
+                            nb_constraints: out_shape.flat_size(),
+                            safe_variance_bound: *max_variance,
+                            variance: var.clone(),
+                        };
+                        let _ = self
+                            .external_variance_constraints
+                            .insert(OperatorIndex(i), constraint);
+                    }
+                }
+            }
+        }
+    }
+
+    fn collect_external_output_constraint(&mut self) {
+        for dag_op in self.dag.get_output_operators_iter() {
+            let DagOperator {
+                id: op_index,
+                shape: out_shape,
+                precision: out_precision,
+                ..
+            } = dag_op;
+            let optional_partition_index = self.partitions.p_cut.partition(&self.dag, op_index);
+            if optional_partition_index.is_none() {
+                continue;
+            }
+            let partition_index = optional_partition_index.unwrap();
+            if self
+                .partitions
+                .p_cut
+                .is_external_partition(&partition_index)
+            {
+                let partitions = self.partitions.clone();
+                let max_variance = partitions.p_cut.external_partitions
```

Comment:
Don't hesitate to name thing to break long expression.
let external_partition_index = p_cut.external_partition_index(partition_index);